paddle/fluid/operators/set_value_op.h

```diff
@@ -0,0 +1,198 @@
+      break;
+    case framework::proto::VarType::BOOL:
+      value_name = "bool_values";
+      break;
+    default:
+      PADDLE_THROW(platform::errors::Unimplemented(
+          "Unsupported data type(code %d) for SetValue operator, only "
+          "supports bool, int32, float32 and int64.",
+          data_type));
+  }
+  return value_name;
+}
+
+inline framework::DDim GetSliceDims(const framework::DDim in_dims,
+                                    const std::vector<int64_t> axes,
+                                    const std::vector<int64_t> starts,
+                                    const std::vector<int64_t> ends) {
+  framework::DDim slice_dims(in_dims);
+
+  for (size_t i = 0; i < axes.size(); ++i) {
+    int64_t axis = axes[i];
+    int64_t dim_value = in_dims[axis];
+
+    int64_t start = starts[i] < 0 ? (starts[i] + dim_value) : starts[i];
+    int64_t end = ends[i] < 0 ? (ends[i] + dim_value) : ends[i];
+    start = std::max(start, static_cast<int64_t>(0));
+    end = std::min(end, dim_value);
+
+    PADDLE_ENFORCE_GT(end, start, platform::errors::InvalidArgument(
+                                      "end should greater than start, but "
+                                      "received end = %d, start = %d",
+                                      end, start));
+    slice_dims[axis] = end - start;
+  }
+  return slice_dims;
+}
+
+template <typename DeviceContext, typename T>
+class SetValueKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const {
+    const int rank = ctx.Output<framework::LoDTensor>("Out")->dims().size();
+    switch (rank) {
+      case 1:
+        SetValueCompute<1>(ctx);
+        break;
+      case 2:
+        SetValueCompute<2>(ctx);
+        break;
+      case 3:
+        SetValueCompute<3>(ctx);
+        break;
+      case 4:
+        SetValueCompute<4>(ctx);
+        break;
+      case 5:
+        SetValueCompute<5>(ctx);
+        break;
+      case 6:
+        SetValueCompute<6>(ctx);
+        break;
+    }
+  }
+
+ private:
+  template <size_t D>
+  void SetValueCompute(const framework::ExecutionContext& ctx) const {
+    auto* in = ctx.Input<framework::LoDTensor>("Input");
+    auto* out = ctx.Output<framework::LoDTensor>("Out");
+
+    auto dtype =
+        static_cast<framework::proto::VarType::Type>(ctx.Attr<int>("dtype"));
+    auto axes = ctx.Attr<std::vector<int64_t>>("axes");
+    auto starts = ctx.Attr<std::vector<int64_t>>("starts");
+    auto ends = ctx.Attr<std::vector<int64_t>>("ends");
+    auto shape = ctx.Attr<std::vector<int64_t>>("shape");
+    auto* value_tensor = ctx.Input<framework::LoDTensor>("ValueTensor");
+
+    auto in_dims = in->dims();
+    auto value_dims = framework::make_ddim(shape);
+    auto slice_dims = GetSliceDims(in_dims, axes, starts, ends);
+
+    auto place = ctx.GetPlace();
+    auto& eigen_place =
+        *ctx.template device_context<DeviceContext>().eigen_device();
+
+    out->ShareDataWith(*in);
```

Comment:
I saw your reply that `in` and `out` are same Paddle.Tensor. Do you mean at program level they are same (i.e. `tensor=set_value(tensor, xxx)`)? Then in my understand, our graph will change it to SSA graph where variable name isn't same (i.e. `tensor1=set_value(tensor0, xxx)`), then the graph doesn't include a circle (you know that our graph doesn't support circle, without SSA, there will be `tensor <--> op` circle)

So at PE and graph level, I guess `out` and `in` are still two different variables. So deleting `in` by memory optimization and `out` loses data may still happen in complex graph.

So I would still suggest one of following:
1. You can copy data, we can make it right and we try to speed up in the future version.
2. Deleting Input variable, then the graph will be complex, such as there will be two ops points to the output in graph:
`op1 -> output <- set_value`
In this case, you have to find a way to handle the running order of set_value is what you want.

For a simple solution, you can do 1.

Reply:
Thanks, as discassed offline, I do 1 now.